### PR TITLE
Complete image-building automation docs

### PR DESCRIPTION
**Pull request by cevich**
This fills out the automation-portion of the processes.  I intend to followup in another PR, with details on manually performing these tasks.  i.e. using the `Makefile` locally.

Signed-off-by: Chris Evich <cevich@redhat.com>

### CODE REVIEW

**Comment by cevich** [comment]
@edsantiago @lsm5  This one is a doozy, but will hopefully be enlightening and informative.  PTAL when you have an afternoon or two, and run through the debugging section.  I've been through this process with @baude but want to ensure the documentation matches and works for you.  ***WARNING:***  Some parts are very network and CPU intensive, and require bare-metal (a VM won't work) :confounded:

Expected outcome: You understand the process (generally) and be able to update base-images, e.g. F32 -> F33 (when it comes out).

Thanks in advance.

**Comment by cevich** [comment]
(rebased / force-pushed)

File: .cirrus.yml [hunk 1]
```diff
@@ -23,6 +23,7 @@ timeout_in: 60m
 task:
     name: "Validate"
     alias: "validate"
+    skip: "changesInclude('**.md', 'LICENSE')"
```
Comment (edsantiago) [inline .cirrus.yml:26]: This raises huge warning flags. As I read this, it means "skip if any .md or LICENSE file was touched". That is not the same as "skip if **ONLY** .md or LICENSE files were touched". But I cannot make any sense of the [cirrus docs](https://cirrus-ci.org/guide/writing-tasks/#supported-functions) nor their example, nor can I find any useful examples on the net, and I don't really want to dive into the Cirrus source code right now.

File: README.md [hunk 2]
```diff
@@ -58,55 +58,143 @@ documentation](https://www.packer.io/docs).
...
-Assuming all you need to do is tweak the package list, add or adjust a makefile
-target, start here.  Before you go deleting all the seemingly "unnecessary" and
-"extra" packages, please remember these VM images are shared by automation
-in multiple repositories :D
+a.k.a. ***Cache Images***
 
-1. VM configuration starts with one of the `cache_images/*_setup.sh` scripts.
+These are the VM Images actually used by other repositories for automated
+testing.  So, assuming you just need to update packages or tweak the list,
+start here.  Though be aware, this repository does not thrroughly test images
```
Comment (edsantiago) [inline README.md:67]: typo, thoroughly

File: README.md [hunk 3]
```diff
@@ -58,55 +58,143 @@ documentation](https://www.packer.io/docs).
...
+in their native-habitat, that's your secondary responsibility (see below).
```
Comment (edsantiago) [inline README.md:68]: I can't figure out what this means ("in their native habitat"), could you rephrase? And can you provide an explicit link such as "see FOO below"?

File: README.md [hunk 4]
```diff
@@ -58,55 +58,143 @@ documentation](https://www.packer.io/docs).
...
-   letter 'c'*, in place of the value next to `_BUILT_IMAGE_SUFFIX:`.  For
-   example, if the url was `http://.../12345` you would paste in `c12345`
-   as the value for `_BUILT_IMAGE_SUFFIX:`.
+5. Go over to whatever other containers/repository needed the image update.
+   Open the `.cirrus.yml` file, and find the 'env' line referencing the image
+   suffix.  It will likely be named `_BUILT_IMAGE_SUFFIX:` or something similar.
 
-7. Open up a PR with this change, and push it.  Once all tests pass and your
+7. Paste in the **big number** *prefixed by the letter 'c'*.  The *"c*" indicates
+   the images are *cache iamges*.  For example, if the url was `http://.../12345`
```
Comment (edsantiago) [inline README.md:109]: typo, images

File: README.md [hunk 5]
```diff
@@ -58,55 +58,143 @@ documentation](https://www.packer.io/docs).
...
+   you would paste in `c12345` as the value for `_BUILT_IMAGE_SUFFIX:`.
+
+8. Open up a PR with this change, and push it.  Once all tests pass and your
```
Comment (edsantiago) [inline README.md:112]: typo, you're

File: README.md [hunk 6]
```diff
@@ -58,55 +58,143 @@ documentation](https://www.packer.io/docs).
...
+9. After all the PRs are merged, you're done.
+
+
+
+
+## The image-builder image (overview step 1)
+
+Google compute engine (GCE) does not provide a wide selection of ready-made
+VM images for use.  Instead, a lengthy and sophisticated process is involved
+to preparing, properly format, and import external VM images for use.  In order
```
Comment (edsantiago) [inline README.md:126]: should be "to prepare"

File: README.md [hunk 7]
```diff
@@ -58,55 +58,143 @@ documentation](https://www.packer.io/docs).
...
+For normal day-to-day use, this process should not need to be modified or
+maintained much.  However, on the off-chance that's ever not true, here is
+an overview of the process followed **by automation** to produce the
+*image-building VM image*:
+
+1. Build the container defined in the `ci` subdirectories `ContainerFile`.
+   Start this container with a copy of the current repository code provided
+   in the `$CIRRUS_WORKING_DIR` directory.
+
+2. Using the `ci` container, execute on the  `make image_builder` target.
```
Comment (edsantiago) [inline README.md:140]: I don't understand this. Could you rephrase or be more explicit?

File: README.md [hunk 8]
```diff
@@ -58,55 +58,143 @@ documentation](https://www.packer.io/docs).
...
+3. The `image_builder/gce.yml` file is converted into JSON format for
+   consumption by the [Hashicorp *packer* utility](https://www.packer.io/).
+
+4. Packer spins up a GCE VM based on CentOS, installs necessary packages
+   and attaches a [nested-virtualization "license" to the
+   VM](https://cloud.google.com/compute/docs/instances/enable-nested-virtualization-vm-instances#enablenestedvirt).
+
+5. The GCE VM is prepared for use, resetting cloud-init metadata, deleting
+   temporary files, ssh host keys, etc.  It is then shut down, while preserving
+   it's disk.
```
Comment (edsantiago) [inline README.md:151]: should be "its", no apostrophe

File: .cirrus.yml (hunk 1 above)
Comment (cevich) [reply .cirrus.yml:26]: oops.  it should be `!changesInclude('**md', 'LICENSE')`

File: README.md (hunk 3 above)
Comment (cevich) [reply README.md:68]: Meaning, it doesn't run podman tests on them.

Comment (cevich) [reply README.md:68]: in fact, right now it doesn't run *any* tests :confused:

File: README.md [hunk 9]
```diff
@@ -120,3 +208,64 @@ Cache-images.  They are then automatically pushed to:
...
+   resources available.  Build the VM Base images by executing the command
+   ``make base_images``. This is the equivilent operation as documented by
+   *overview step 2*.  ***N/B*** The GCS -> GCE image conversion can take
+   some time, be patient.  Packer may not produce any output for several minutes
+   while the conversion is happening.
+
+5. When successful, the names of the produced images will all be referenced
+   in the `base_images/manifest.json` file.  If there are problems, fix them
+   and remove the `manifest.json` file.  Then re-run the same *make* command
+   as before, packer it will force-overwrite any broken/partially created
```
Comment (edsantiago) [inline README.md:260]: spurious "it"?

File: README.md [hunk 10]
```diff
@@ -120,3 +208,64 @@ Cache-images.  They are then automatically pushed to:
...
+   ``make image_builder_debug GAC_FILEPATH=</home/path/to/gac.json> IMG_SFX=<UUID chosen in step 1>``
+
+3. You will be dropped into a debugging container, inside a volume-mount of
+   the repository root.  This container is practically identical to the VM
+   produced and used in *overview step 1*.  If changes are made, the container
+   image should be re-built to reflect them.
+
+4. Still within the container, again ensure you have plenty of network and CPU
+   resources available.  Build the VM Base images by executing the command
+   ``make base_images``. This is the equivilent operation as documented by
```
Comment (edsantiago) [inline README.md:252]: typo, equivalent

File: README.md [hunk 11]
```diff
@@ -120,3 +208,64 @@ Cache-images.  They are then automatically pushed to:
...
+   some time, be patient.  Packer may not produce any output for several minutes
+   while the conversion is happening.
+
+5. When successful, the names of the produced images will all be referenced
+   in the `base_images/manifest.json` file.  If there are problems, fix them
+   and remove the `manifest.json` file.  Then re-run the same *make* command
+   as before, packer it will force-overwrite any broken/partially created
+   images automatically.
+
+6. Produce the GCE VM Cache Images, equivilent to the operations outlined
```
Comment (edsantiago) [inline README.md:263]: typo, equivalent

File: README.md (hunk 7 above)
Comment (cevich) [reply README.md:140]: Meaning the container from step 1.  The doc. assumes you know how to `-v $PWD:$PWD:Z -w $PWD`.  Maybe I should put an example as a reminder?

File: README.md [hunk 12]
```diff
@@ -58,55 +58,143 @@ documentation](https://www.packer.io/docs).
...
+to perform these steps within automation, a dedicated VM image is needed which
+itself has been prepared with the necessary incantations, packages, configuration,
+and magic license keys.
+
+For normal day-to-day use, this process should not need to be modified or
+maintained much.  However, on the off-chance that's ever not true, here is
+an overview of the process followed **by automation** to produce the
+*image-building VM image*:
+
+1. Build the container defined in the `ci` subdirectories `ContainerFile`.
```
Comment (edsantiago) [inline README.md:136]: Oh okay now I understand this. Typo confused me: should be "subdirectory's"

File: README.md (hunk 7 above)
Comment (edsantiago) [reply README.md:140]: ...and yes, an actual example is always helpful.

File: README.md (hunk 12 above)
Comment (cevich) [reply README.md:136]: oops, yes, you are correct.

File: README.md (hunk 3 above)
Comment (edsantiago) [reply README.md:68]: Have you double-checked your YAML? The `<<:` thing inside `matrix` in `base_images` and `cache_images` does not seem to do what you think it does. (Discovered because I got way too lost reviewing this, and had to convert the cirrus.yml to actual-yaml to use my graph tool, which then drew nonsensical results, which turned out to be due to the `<<` things)

File: README.md (hunk 8 above)
Comment (cevich) [reply README.md:151]: sigh...my brainbits never can seem to get that one right :confused:

**Comment by cevich** [comment]
Rebased/force pushed w/ all recommended changes.

File: README.md [hunk 13]
```diff
@@ -58,55 +58,152 @@ documentation](https://www.packer.io/docs).
...
+2. From within the `ci` container (above), in the repository root volume, execute
+   the  `make image_builder` target.
+
+3. The `image_builder/gce.yml` file is converted into JSON format for
+   consumption by the [Hashicorp *packer* utility](https://www.packer.io/).
+   This generated file may be ignored, *make* will be regenerate it upon
+   any changes to the YAML file.
+
+4. Packer will spin up a GCE VM based on CentOS, installs necessary packages
+   and attaches a [nested-virtualization "license" to the
```
Comment (edsantiago) [inline README.md:153]: The change here now leaves "installs" and "attaches" disagreeing with "will spin up"; but this isn't worth re-pushing for

File: README.md [hunk 14]
```diff
@@ -58,55 +58,152 @@ documentation](https://www.packer.io/docs).
...
+   VM](https://cloud.google.com/compute/docs/instances/enable-nested-virtualization-vm-instances#enablenestedvirt).  Be patient until this process completes.
 
+5. Near the end of the process, the GCE VM delete temporary files, ssh host keys,
```
Comment (edsantiago) [inline README.md:156]: Similarly, this should probably be "**will** delete ..."

File: README.md [hunk 15]
```diff
@@ -58,55 +58,152 @@ documentation](https://www.packer.io/docs).
...
+   etc. and then shut down.
 
-## Container images
+6. Packer should then automatically call into the google cloud APIs and
+   coordinate conversion of the VM disk into a bootable image.  Again,
+   please be patient for this process to complete, it may take several minutes.
+
+7. When finished, packer will writes the freshly created image name and other
```
Comment (edsantiago) [inline README.md:163]: All these changes! Should be "will **write**", not writes

File: .cirrus.yml (hunk 1 above)
Comment (cevich) [reply .cirrus.yml]: Hmmm, nope, inverting the condition doesn't work as the docs suggest :confused:

Okay I have an idea...

File: README.md (hunk 13 above)
Comment (cevich) [reply README.md:153]: Nope, I got it.

File: README.md (hunk 15 above)
Comment (cevich) [reply README.md:163]: they're not changes, they're FIXES :grin:

File: .cirrus.yml [hunk 16]
```diff
@@ -23,6 +23,16 @@ timeout_in: 60m
...
+    # Ref: https://cirrus-ci.org/guide/writing-tasks/#supported-functions
+    skip: &notdocs <-
+        !changesInclude(
+          'base_images/***',
+          'cache_images/***',
+          'ci/***',
+          'image_builder/***',
+          'podman/***',
+          '*.sh',
+          'Makefile')
```
Comment (edsantiago) [inline .cirrus.yml:35]: ...and `.cirrus.yml`.

I really don't think this is a wise approach; I would try filing a Cirrus PR asking sweetly for `changesAreOnly(...)`

**Comment by cevich** [comment]
@edsantiago ooof okay I'm done updating this for now.  I don't think I can get the `skip` thing working properly.  May have to forget that feature for now or fall back to interpreting magic strings in the title (not my fav.).

File: .cirrus.yml (hunk 16 above)
Comment (cevich) [reply .cirrus.yml:35]: ya, this would be a PITA to maintain, even if it worked.

Comment (cevich) [reply .cirrus.yml:35]: TBH, it's probably a decision best left up to humans.  I've never had good experiences with these kinds of execution-conditions.  There always seems to be some corner-case or another that breaks it.  Whereas just sticking `[CI:DOCS]` or similar in the title, both functions as directed and leaves a record of what happened (in the logs).

**Comment by edsantiago** [comment]
I'm totally OK with leaving the `skip` as future work.

**Comment by cevich** [comment]
> I'm totally OK with leaving the skip as future work.

My thinking also, it's easy enough to just press cancel for now.

File: README.md [hunk 17]
```diff
@@ -120,3 +217,64 @@ Cache-images.  They are then automatically pushed to:
...
+
+1. Invent some unique identity suffix for your images.  It may contain (***only***)
+   lowercase letters, numbers and dashes; nothing else.  Some suggestions
+   of useful values would be your name and todays date.  If you manage to screw
+   this up somehow, stern errors will be presented without causing any real harm.
+
+2. Ensure you have podman installed, and lots of available network and CPU
+   resources (i.e. turn off YouTube, shut down background VMs and other hungry
+   tasks).  Build the image-builder container image, by executing
+   ``make image_builder_debug GAC_FILEPATH=</home/path/to/gac.json> IMG_SFX=<UUID chosen in step 1>``
```
Comment (lsm5) [inline README.md:252]: @cevich this step ends for me with:

```
Error: host directory cannot be empty
make: *** [Makefile:164: image_builder_debug] Error 125
```

Whatsup with that?

**Review by lsm5 — APPROVED** [review]
I walked through the debugging section. 

It told me disk images were created and also that artifacts were exported.

So, LGTM from me.

**Comment by cevich** [comment]
Fantastic @lsm5 thanks a bunch for going through that.  It gives me confidence the documentation is on-target.

**Comment by cevich** [comment]
@edsantiago okay if I merge this or would you like to take a manual run through the instructions?  IMHO, it's not necessary as both Lokesh and Brent have been through them successfully.

**Comment by edsantiago** [comment]
@lsm5's and @baude's experience is good enough for me!

**Comment by cevich** [comment]
thanks @edsantiago and @lsm5 and @baude
